src/index/src/inverted_index/create/sort/intermediate_rw/codec_v1.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io;
+
+use bytes::{Buf, BufMut};
+use common_base::BitVec;
+use futures_codec::{BytesMut, Decoder, Encoder};
+use snafu::{location, Location};
+
+use crate::inverted_index::error::{Error, Result};
+use crate::inverted_index::Bytes;
+
+const U64_LENGTH: usize = std::mem::size_of::<u64>();
+
+/// Magic bytes for this intermediate codec version
+pub const CODEC_V1_MAGIC: &[u8; 4] = b"im01";
+
+/// Codec for serializing and deserializing intermediate data for external sorting.
+///
+/// Binary format serialization. The item is laid out as follows:
+/// ```text
+/// [value len][value][bitmap len][bitmap]
+///     [8]     [?]       [8]       [?]
+/// ```
+pub struct IntermediateCodecV1;
+
+/// [`FramedWrite`] requires the [`Encoder`] trait to be implemented.
+impl Encoder for IntermediateCodecV1 {
+    type Item = (Bytes, BitVec);
+    type Error = Error;
+
+    fn encode(&mut self, item: (Bytes, BitVec), dst: &mut BytesMut) -> Result<()> {
+        let value_bytes = item.0;
+        let bitmap_bytes = item.1.into_vec();
+
+        dst.reserve(U64_LENGTH * 2 + value_bytes.len() + bitmap_bytes.len());
+        dst.put_u64_le(value_bytes.len() as u64);
+        dst.extend_from_slice(&value_bytes);
+        dst.put_u64_le(bitmap_bytes.len() as u64);
+        dst.extend_from_slice(&bitmap_bytes);
+        Ok(())
+    }
+}
+
+/// [`FramedRead`] requires the [`Decoder`] trait to be implemented.
+impl Decoder for IntermediateCodecV1 {
+    type Item = (Bytes, BitVec);
+    type Error = Error;
+
+    /// Decodes the `src` into `(Bytes, BitVec)`. Returns `None` if
+    /// the `src` does not contain enough data for a complete item.
+    ///
+    /// Only after successful decoding, the `src` is advanced. Otherwise,
+    /// it is left untouched to wait for filling more data and retrying.
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>> {
+        // [value len][value][bitmap len][bitmap]
+        //     [8]     [?]       [8]       [?]
+
+        // decode value len
+        if src.len() < U64_LENGTH {
+            return Ok(None);
+        }
+        let (value_len, buf) = src.split_at(U64_LENGTH);
+        let value_len = u64::from_le_bytes(value_len.try_into().unwrap()) as usize;
+
+        // decode value
+        if buf.len() < value_len {
+            return Ok(None);
```

Comment:
Is such a case an error or normal? It's possible when writing fails, but I am not sure if it's ok to return `None`.

Reply:
It's a framed read. `FramedReader` may read up to the end of a frame (not enough remaining bytes for one Item), in which case `FramedReader` needs to continue adding bytes and retry. Only when there are no subsequent bytes will `FramedReader` raise an error.

Reply:
When writing fails, it eventually returns an error thrown by the reader.